File: src/jvmTest/java/org/hisp/dhis/lib/expression/util/TNode.java
package org.hisp.dhis.lib.expression.util;

import lombok.Value;
import org.hisp.dhis.lib.expression.ast.Node;
import org.hisp.dhis.lib.expression.ast.NodeType;

import java.util.ArrayList;
import java.util.List;
import java.util.function.UnaryOperator;

/**
 * A simplified tree structure useful to verify the {@link NodeType} structure a {@link Node} AST.
 *
 * @author Jan Bernitt
 */
@Value(staticConstructor = "of")
public class TNode {

    NodeType type;
    List<TNode> children;

    public static TNode ofLevels(NodeType root, NodeType... level1toN) {
        TNode node = TNode.of(root);
        TNode target = node;
        for (NodeType t : level1toN) {
            TNode c = of(t);
            target.children.add(c);
            target = c;
        }
        return node;
    }

    public static TNode of(NodeType type) {
        return new TNode(type, new ArrayList<>());
    }

    public static TNode of(Node<?> root) {
        return root.map(Node::getType, TNode::new);
    }

    public TNode add(NodeType type) {
        children.add(of(type));
        return this;
    }

    public TNode add(NodeType type, UnaryOperator<TNode> with) {
        children.add(with.apply(of(type)));
        return this;
    }

    public TNode add(NodeType type, NodeType... level1toN) {
        children.add(ofLevels(type, level1toN));
        return this;
    }
}

File: src/jvmTest/java/org/hisp/dhis/lib/expression/util/TNode.kt
package org.hisp.dhis.lib.expression.util

import org.hisp.dhis.lib.expression.ast.Node
import org.hisp.dhis.lib.expression.ast.NodeType
import java.util.function.UnaryOperator

/**
 * A simplified tree structure useful to verify the [NodeType] structure of a [Node] AST.
 *
 * @author Jan Bernitt
 */
data class TNode(
    val type: NodeType,
    val children: MutableList<TNode>
) {

    fun add(type: NodeType): TNode {
        children.add(of(type))
        return this
    }

    fun add(type: NodeType, with: UnaryOperator<TNode>): TNode {
        children.add(with.apply(of(type)))
        return this
    }

    fun add(type: NodeType, vararg level1toN: NodeType): TNode {
        children.add(ofLevels(type, *level1toN))
        return this
    }

    companion object {
        fun ofLevels(root: NodeType, vararg level1toN: NodeType): TNode {
            val node = of(root)
            var target = node
            for (t in level1toN) {
                val c = of(t)
                target.children.add(c)
                target = c
            }
            return node
        }

        fun of(type: NodeType): TNode {
            return TNode(type, ArrayList())
        }

        fun of(root: Node<*>): TNode {
            return root.map(Node<*>::getType, ::TNode)
        }
    }
}
